src/uct/tcp/tcp_cm.c

```diff
@@ -8,100 +8,289 @@
+            uct_tcp_iface_outstanding_inc(iface);
+        } else {
+            ucs_assert(old_conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_CONNECTED:
+        ucs_assert((old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) ||
+                   (old_conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING));
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_CLOSED:
+        ucs_assert(old_conn_state != UCT_TCP_EP_CONN_STATE_CLOSED);
+        if ((old_conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) ||
+            (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK)) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    default:
+        ucs_assert(ep->conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING);
+        break;
+    }
+
+    ucs_debug("tcp_ep %p: %s -> %s for the [%s]<->[%s] connection",
+              ep, uct_tcp_ep_cm_state[old_conn_state].name,
+              uct_tcp_ep_cm_state[ep->conn_state].name,
+              ucs_sockaddr_str((const struct sockaddr*)&iface->config.ifaddr,
+                               str_local_addr, UCS_SOCKADDR_STRING_LEN),
+              ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                               UCS_SOCKADDR_STRING_LEN));
+}
+
+static ucs_status_t uct_tcp_cm_send_conn_req(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface         = ucs_derived_of(ep->super.super.iface,
+                                                    uct_tcp_iface_t);
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_REQ
+    };
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    size_t addr_size;
+    ucs_status_t status;
+
+    status = ucs_sockaddr_sizeof((const struct sockaddr*)&iface->config.ifaddr,
+                                 &addr_size);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    memcpy(&conn_pkt.iface_addr, &iface->config.ifaddr, addr_size);
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_debug("tcp_ep %p: conection request sent to %s",
+              ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_recv_conn_req(uct_tcp_ep_t *ep,
+                                             struct sockaddr *peer_addr)
+{
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    uct_tcp_cm_conn_pkt_t conn_pkt;
+    ucs_status_t status;
+
+    status = ucs_socket_recv(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_assertv(conn_pkt.event == UCT_TCP_CM_CONN_REQ, "ep=%p", ep);
```

Comment:
need to have ucs_error here, since its some wrong data received from the network, not local problem necessarily
same for uct_tcp_cm_recv_conn_ack